### PR TITLE
filter out not accessible members

### PR DESCRIPTION
**Pull request by mwielocha**
Request for #1198 
I have a version of this for both master and #1412 branch.
Will merge when `no netbeans` gets approved.

### CODE REVIEW

**Comment by fommil** [comment]
this is great stuff, I really noticed the private stuff when I was doing my recent Java hacking. There is a lot of stuff to filter out, it would appear :smile: (and almost everything I wanted access to turned out to be `private`... bloody stdlib...)

File: core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala [hunk 1]
```diff
@@ -208,21 +208,26 @@ trait JavaCompletion extends Helpers with SLF4JLogging {
...
+
+        info.getElements.getAllMembers(tel).filter(isAccessible).flatMap { el =>
+          filterElement(info, el, prefix, caseSense, importing, false)
         }
-      }
-    }.getOrElse(List())
-    candidates.toList
+
+      case e =>
+        log.warn("Unrecognized type element " + e)
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: you might want to create a ticket to capture these, with detailed instructions of what you want people to do, and then  direct people to the ticket? You never know, somebody might add support for more trees.

**Comment by mwielocha** [comment]
Yup, btw. the filter out `<init>` is for the `no netbeans` stuff, I missed that because of all the noise from private members.

File: core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala [hunk 2]
```diff
@@ -208,21 +208,26 @@ trait JavaCompletion extends Helpers with SLF4JLogging {
...
+        List.empty
+
+    }.map(_.toList).getOrElse(List.empty)
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: do you know about `breakOut`? Highly recommended watching: https://skillsmatter.com/skillscasts/6939-optimising-scala-for-fun-and-profit (you'll never call `.toList`, `.toMap` or `.toSet` again...)

File: core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala [hunk 3]
```diff
@@ -208,21 +208,26 @@ trait JavaCompletion extends Helpers with SLF4JLogging {
...
   ): List[CompletionInfo] = {
-    val candidates = typeElement(info, target).map { el =>
-      el match {
-        case tel: TypeElement => {
-          val elements: Elements = info.getElements()
-          elements.getAllMembers(tel).flatMap { e =>
-            filterElement(info, e, prefix, caseSense, importing, false)
-          }
+
+    typeElement(info, target).map {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: is this maybe better as a `flatMap`? I realise this is just code you're changing but no reason not to clean it up at the same time.

Comment (mwielocha) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: I am now a bit more cautious when cleaning up, after the pattern match from  #1412 ;P

Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: You'll soon learn to put one contentious style change in every PR, to distract us from the really hard bits :smile:

**Comment by mwielocha** [comment]
So what is going on here? Isn't this auto-scalariformatted?

**Comment by fommil** [comment]
scalariform is a bit weird, it formats source code as a pre compile phase, so you probably made changes without compiling.

File: core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala [hunk 4]
```diff
@@ -209,7 +210,7 @@ trait JavaCompletion extends Helpers with SLF4JLogging {
     caseSense: Boolean
   ): List[CompletionInfo] = {
 
-    typeElement(info, target).map {
+    typeElement(info, target).toList.flatMap {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: there shouldn't be any need for the `toList` here

Comment (mwielocha) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: I must admit, there is something here that I am missing: so you're saying that this: `Option(1).flatMap(x => List(x))` should compile, right?

Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: yes, I think so (unless this is a gotcha)

Comment (mwielocha) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: Not a gotcha, I was honestly wondering how can this be done nicely, but unfortunately this only works one way: `List(1).flatMap(x => Option(x))`, either way I'm pretty sure that `Option.toList` is super-cheap, so shouldn't really be a problem here.

**Comment by mwielocha** [comment]
Not sure if this: `java.lang.AssertionError: assertion failed: timeout (10 minutes) during expectMsgClass waiting for interface` is in any way connected with this pr...

File: core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala (hunk 4 above)
Comment (fommil) [inline core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala]: oh sure, if it's on an Option then it's fine. I couldn't see the starting type.

**Comment by fommil** [comment]
if it's in the `DebugTest` then it's just an annoying transient we've not tracked down yet.

**Comment by fommil** [comment]
well I think this is a lot less contentious than the Netbeans removal, so I'll just accept this as is. Thank you!

**Pull request merged**